pkg/asset/machines/gcp/gcpmachines.go

```diff
@@ -0,0 +1,112 @@
+// Package gcp generates Machine objects for gcp.
+package gcp
+
+import (
+	"fmt"
+
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/utils/ptr"
+	capg "sigs.k8s.io/cluster-api-provider-gcp/api/v1beta1"
+	capi "sigs.k8s.io/cluster-api/api/v1beta1"
+
+	machineapi "github.com/openshift/api/machine/v1beta1"
+	"github.com/openshift/installer/pkg/asset"
+	"github.com/openshift/installer/pkg/types"
+	gcptypes "github.com/openshift/installer/pkg/types/gcp"
+)
+
+// GenerateMachines returns manifests and runtime objects to provision nodes using CAPI.
+func GenerateMachines(clusterID string, region string, subnet string, pool *types.MachinePool, role string, projectID string, userTags map[string]string, userLabels map[string]string) ([]*asset.RuntimeFile, error) {
+	if poolPlatform := pool.Platform.Name(); poolPlatform != gcptypes.Name {
+		return nil, fmt.Errorf("non-GCP machine-pool: %q", poolPlatform)
+	}
+	mpool := pool.Platform.GCP
+
+	total := int64(1)
+	if pool.Replicas != nil {
+		total = *pool.Replicas
+	}
+
+	var osImage string
+	if mpool.OSImage != nil {
+		osImage = mpool.OSImage.Name
+	}
+
+	// Set SecureBoot
+	shieldedInstanceConfig := capg.GCPShieldedInstanceConfig{}
+	if mpool.SecureBoot == string(machineapi.SecureBootPolicyEnabled) {
+		shieldedInstanceConfig.SecureBoot = capg.SecureBootPolicyEnabled
+	}
+
+	// Set ServiceAccount
+	instanceServiceAccount := fmt.Sprintf("%s-%s@%s.iam.gserviceaccount.com", clusterID, role[0:1], projectID)
+	// TODO - create a different service account for an xpn installation similar to current implementation ?
+
+	objectLabels := make(map[string]string, 1)
+	if role == "master" {
+		objectLabels["cluster.x-k8s.io/control-plane"] = ""
+	} else {
+		objectLabels["cluster.x-k8s.io/worker"] = ""
+	}
+
+	// Create CAPG machine
+	var result []*asset.RuntimeFile
+	for idx := int64(0); idx < total; idx++ {
+		gcpMachine := &capg.GCPMachine{
+			TypeMeta: metav1.TypeMeta{
+				APIVersion: "infrastructure.cluster.x-k8s.io/v1beta1",
+				Kind:       "GCPMachine",
+			},
+			ObjectMeta: metav1.ObjectMeta{
+				Name:   fmt.Sprintf("%s-%s-%d", clusterID, pool.Name, idx),
+				Labels: objectLabels,
+			},
+			Spec: capg.GCPMachineSpec{
+				InstanceType:           mpool.InstanceType,
+				Subnet:                 ptr.To(subnet),
+				AdditionalLabels:       userLabels,
+				Image:                  ptr.To(osImage),
+				RootDeviceType:         ptr.To(capg.DiskType(mpool.OSDisk.DiskType)),
+				RootDeviceSize:         mpool.OSDisk.DiskSizeGB,
+				ShieldedInstanceConfig: ptr.To(shieldedInstanceConfig),
+				ServiceAccount: &capg.ServiceAccount{
+					Email:  instanceServiceAccount,
+					Scopes: []string{"https://www.googleapis.com/auth/cloud-platform"},
```

Comment:
how did you determine this scope? if it is documented clearly (big if) maybe put a link here in a comment, perhaps set a variable for this `&capg.ServiceAccount` earlier, because we will need to pipe in the value from the install config.

Reply:
> how did you determine this scope? if it is documented clearly (big if) maybe put a link here in a comment, perhaps set a variable for this `&capg.ServiceAccount` earlier, because we will need to pipe in the value from the install config.

That was a copy/paste from https://github.com/openshift/installer/blob/master/pkg/asset/machines/gcp/machines.go#L215
I'm not quite clear where it was derived from, I'll try to figure it out to add a comment.
Will move this setting to earlier when the email is set.

Reply:
So it seems that the scope is documented here
https://cloud.google.com/compute/docs/access/create-enable-service-accounts-for-instances
"Set the cloud platform (https://www.googleapis.com/auth/cloud-platform) scope on your VM."